text/0000-ptr-meta.md

```diff
@@ -0,0 +1,320 @@
+}
+
+/// The vtable for a trait object.
+///
+/// A vtable (virtual call table) represents all the necessary information
+/// to manipulate the concrete type stored inside a trait object.
+/// It notably it contains:
+///
+/// * type size
+/// * type alignment
+/// * a pointer to the type’s `drop_in_place` impl (may be a no-op for plain-old-data)
+/// * pointers to all the methods for the type’s implementation of the trait
+///
+/// Note that the first three are special because they’re necessary to allocate, drop,
+/// and deallocate any trait object.
+///
+/// The layout of vtables is still unspecified, so this type is one a more-type-safe
+/// convenience for accessing those 3 special values. Note however that `VTable` does
+/// not actually know the trait it’s associated with, indicating that, at very least,
+/// the location of `size`, `align`, and `drop_in_place` is identical for all
+/// trait object vtables in a single program.
+pub struct VTable {
+    _priv: (),
+}
+
+impl VTable {
+    /// Returns the size of the type associated with this vtable.
+    pub fn size(&self) -> usize { ... }
+
+    /// Returns the alignment of the type associated with this vtable.
+    pub fn align(&self) -> usize { ... }
+
+    /// Returns the size and alignment together as a `Layout`
+    pub fn layout(&self) -> alloc::Layout {
+        unsafe {
+            alloc::Layout::from_size_align_unchecked(self.size(), self.align())
+        }
+    }
+
+    /// Drops the value pointed at by `data` assuming it has the type
+    /// associated with this vtable.
+    ///
+    /// Behaviour is Undefined if the type isn’t correct or the pointer
+    /// is invalid to reinterpret as `&mut TheType`.
+    pub unsafe fn drop_in_place(&self, data: *mut ()) { ... }
+}
+```
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The status quo is that code (such as linked in [Motivation]) that requires this functionality
+needs to transmute to and from `std::raw::TraitObject`
+or a copy of it (to be compatible with Stable Rust).
+Additionally, in cases where constructing the data pointer
+requires knowing the alignment of the concrete type,
+a dangling pointer such as `0x8000_0000_usize as *mut ()` needs to be created.
+It is not clear whether `std::mem::align_of(&*ptr)` with `ptr: *const dyn SomeTrait`
+is Undefined Behavior with a dangling data pointer.
+
+A [previous iteration][2579] of this RFC proposed a `DynTrait`
+that would only be implemented for trait objects like `dyn SomeTrait`.
+There would be no `Metadata` associated type, `&'static VTable` was hard-coded in the trait.
+In addition to being more general
+and (hopefully) more compatible with future custom DSTs proposals,
+this RFC resolves the question of what happens
+if trait objects with super-fat pointers with multiple vtable pointers are ever added.
+(Answer: they can use a different metadata type like `[&'static VTable; N]`.)
```

Comment:
Should then `[&'static VTable; 1]` for `dyn SomeTrait` be used to make that transition smoother and to fit better with const generics?

Reply:
This would make some sense if we were definitely gonna have super-fat pointers with multiple separate vtable pointers as fat pointer metadata. But if we don’t and end up with a different solution to upcasting, we’ll end up with a always-single-item arrays for no reason. This isn’t really the thread to get into that discussion, but my opinion is that super-fat pointer have a significant enough size cost that I’d much prefer a different solution.

Perhaps an alternative for this RFC, more neutral with respect super-fat pointers v.s. not, would be to have `type Metadata = VTable<Self>;` for trait objects. (See other comments about `VTable`’s possible type paramater.) With the pointer/reference indirection hidden away in private fields of the `VTable` type, this design would be compatible with having `VTable<dyn A + B>` contain two pointers in the future.

Reply:
Also, is there a use case for generic code that accepts any trait object with any number of vtable pointer but not other kinds of DSTs?